src/controllers/rankings.js

```diff
@@ -141,106 +177,54 @@ async function computeTopRankings(numTopRankings, allRankings) {
+const getTopRankings = establishList(list => async (ctx, next) => {
   const num = numFromCtx(ctx)
   const depth = depthFromCtx(ctx)
-  const allRankings = await getAllUserRankings(depth)
+  const allRankings = await allUserRankingsForList(list)(depth)
 
   ctx.body = await computeTopRankings(num, allRankings)
 
   await next()
-}
-
-function generateRankingsRecursive(n) {
-  if (n === 1) {
-    return 'nextRanking { ...RankingFields }'
-  }
-  return `nextRanking { ...RankingFields ${generateRankingsRecursive(n - 1)}}`
-}
+})
 
-function flattenRankings(currentItem) {
-  if (!currentItem.nextRanking) {
-    return [currentItem.item]
-  } else {
-    return [currentItem.item, ...flattenRankings(currentItem.nextRanking)]
-  }
-}
-
-async function createUsers(ctx, next) {
-  const userRankings = {
-    // 1: [5,1,4,5,2],
-    // 2: [2,3,1,5,4],
-    // 3: [3,2,1,4,5],
-    // 4: [2,1,3,4,5],
-    // 5: [5,2,1,4,3],
-    // 6: [1,4,5,3,2]
-    1: [5,1,4,10,2,8],
-    2: [2,3,1,5,4,7,8,6,10,9],
-    3: [3,2,1,4,5],
-    4: [2],
-    5: [5,2,1,4,3],
-    6: [1,4,8,4,9,10,6,7,5,3,2],
-    7: [5,3,2,1,6,7,8,4,9,10]
-  }
-
-  const asyncForEach = async (f, data) => {
-    if (data.length === 0) return;
-    await f(R.head(data))
-    await asyncForEach(f, R.tail(data))
-  }
-
-  await asyncForEach(async ([user_id, itemIds]) => {
-    user_id = +user_id
-    let prevUserRankingId = null
-    await asyncForEach(async item_id => {
-      item_id = +item_id
-
-      const newRanking = await ctx.app.models.UserRanking.insertAfter(prevUserRankingId, {
-        item_id,
-        user_id
-      })
-      console.log('inserted', user_id, item_id)
-      prevUserRankingId = newRanking.id
-    }, itemIds)
-  }, R.toPairs(userRankings))
-
-  await next()
-}
-
-async function clearUserRankings(ctx, next) {
-  await saveSnapshot(ctx);
-
-  const deletedUserRankings = await ctx.app.models.UserRanking.query().delete();
-  ctx.body = deletedUserRankings;
-
-  await next()
-}
-
-async function saveSnapshot(ctx) {
+const saveSnapshot = establishList(list => async (ctx, next) => {
   const {
     UserRanking, RankingsSnapshot
   } = ctx.app.models
 
   const num = numFromCtx(ctx)
   const depth = depthFromCtx(ctx)
-  const allUserRankingsLists = await getAllUserRankings(depth)
+  const allUserRankingsLists = await allUserRankingsForList(list)(depth)
 
   const user_rankings = await UserRanking.query()
+    .where({list_id: list.id})
   const top_rankings = await computeTopRankings(num, allUserRankingsLists)
 
   await RankingsSnapshot.query()
     .insert({
       snapshot: {
+        list,
         user_rankings,
         top_rankings
       }
     })
-}
+})
+
+const clearUserRankings = establishList(list => async (ctx, next) => {
+  await saveSnapshot(ctx);
+
+  const deletedUserRankings = await ctx.app.models.UserRanking.query()
+    .delete()
+    .where({list_id: list.id})
+
+  ctx.body = deletedUserRankings;
+
+  await next()
+})
```

Comment:
You'll notice there are 0 `function` keywords in this file.

![](https://media.giphy.com/media/3oD3YJVnCAtIY050ha/giphy.gif)